circuits/cpp/src/aztec3/circuits/rollup/base/native_base_rollup_circuit.cpp

```diff
@@ -428,15 +431,107 @@ AppendOnlySnapshot check_nullifier_tree_non_membership_and_insert_to_tree(DummyC
     };
 }
 
-AppendOnlySnapshot insert_state_transitions(BaseRollupInputs const& baseRollupInputs)
+fr insert_state_transitions(
+    DummyComposer& composer,
+    fr tree_root,
+    std::array<abis::PublicDataWrite<NT>, STATE_TRANSITIONS_LENGTH> const& state_transitions,
+    size_t witnesses_offset,
+    std::array<abis::MembershipWitness<NT, PUBLIC_DATA_TREE_HEIGHT>, 2 * STATE_TRANSITIONS_LENGTH> const& witnesses)
 {
-    // TODO: Implement me
-    auto root = baseRollupInputs.start_public_data_tree_snapshot.root;
+    auto root = tree_root;
 
-    return {
-        .root = root,
-        .next_available_leaf_index = 0,
-    };
+    for (size_t i = 0; i < STATE_TRANSITIONS_LENGTH; ++i) {
+        const auto& state_write = state_transitions[i];
+        const auto& witness = witnesses[i + witnesses_offset];
+
+        if (state_write.is_empty())
+            continue;
+
+        composer.do_assert(
+            witness.leaf_index == state_write.leaf_index,
+            format("mismatch state write ", state_write.leaf_index, " and witness leaf index ", witness.leaf_index),
+            CircuitErrorCode::BASE__INVALID_PUBLIC_READS);
+
+        check_membership<NT>(composer,
+                             state_write.old_value,
+                             state_write.leaf_index,
+                             witness.sibling_path,
+                             root,
+                             format("validate_state_reads index ", i));
+
+        root = root_from_sibling_path<NT>(state_write.new_value, state_write.leaf_index, witness.sibling_path);
+    }
+
+    return root;
+}
+
+void validate_state_reads(
+    DummyComposer& composer,
+    fr tree_root,
+    std::array<abis::PublicDataRead<NT>, STATE_READS_LENGTH> const& state_reads,
+    size_t witnesses_offset,
+    std::array<abis::MembershipWitness<NT, PUBLIC_DATA_TREE_HEIGHT>, 2 * STATE_READS_LENGTH> const& witnesses)
+{
+    for (size_t i = 0; i < STATE_READS_LENGTH; ++i) {
+        const auto& state_read = state_reads[i];
+        const auto& witness = witnesses[i + witnesses_offset];
+
+        if (state_read.is_empty())
+            continue;
+
+        composer.do_assert(
+            witness.leaf_index == state_read.leaf_index,
+            format("mismatch state read ", state_read.leaf_index, " and witness leaf index ", witness.leaf_index),
+            CircuitErrorCode::BASE__INVALID_PUBLIC_READS);
+
+        info(" Checking membership for ",
```

Comment:
is this meant to stay, it makes the test output extremely noisy

Reply:
Definitely not, thanks for the catch!